### ISSUE TITLE
FP on rule S3801 with switch over union types

### ISSUE THREAD

**Issue by m-g-sonar** [SonarSource/SonarTS#702]
Current implementation of [RSPEC-3801](https://jira.sonarsource.com/browse/RSPEC-3801) should cover [union types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types) when using switch, as described in specification about [exhaustiveness checking](https://www.typescriptlang.org/docs/handbook/advanced-types.html#exhaustiveness-checking).

For instance:
```typescript
type Shape = Square | Rectangle | Circle | Triangle;

// all the case are covered from type Shape, so S3801 should not raise an issue
function area(s: Shape): Shape {
    switch (s.kind) {
        case "square": 
          return ...
        case "rectangle": 
          return ...
        case "circle": 
          return ... 
        case "triangle"
          return ...
    }
}
```
